lib/src/menus/popovers.dart

```diff
@@ -0,0 +1,536 @@
+/// The popover is automatically closed when the user taps outside of its bounds.
+///
+/// Provide a [popoverGeometry] to control the size and position of the popover. The popover
+/// is first sized given the [PopoverGeometry.constraints] and then positioned using the
+/// [PopoverGeometry.aligner].
+///
+/// When the popover is displayed it requests focus to itself, so the user can
+/// interact with the content using the keyboard.
+class PopoverScaffold extends StatefulWidget {
+  const PopoverScaffold({
+    super.key,
+    required this.controller,
+    required this.buttonBuilder,
+    required this.popoverBuilder,
+    this.popoverGeometry = const PopoverGeometry(),
+    this.popoverFocusNode,
+    this.parentFocusNode,
+    this.boundaryKey,
+    this.onTapOutside = _PopoverScaffoldState.closePopoverOnTapOutside,
+  });
+
+  /// Shows and hides the popover.
+  final PopoverController controller;
+
+  /// Builds a button that is always displayed.
+  final WidgetBuilder buttonBuilder;
+
+  /// Builds the content of the popover.
+  final WidgetBuilder popoverBuilder;
+
+  /// Controls the size and position of the popover.
+  ///
+  /// The popover is first sized, then positioned.
+  final PopoverGeometry popoverGeometry;
+
+  /// The [FocusNode] which is bound to the popover.
+  ///
+  /// Focus will be requested to this [FocusNode] when the popover is displayed.
+  final FocusNode? popoverFocusNode;
+
+  /// The [FocusNode], to which the popover [FocusNode] will be added as a child.
+  ///
+  /// In Flutter, [FocusNode]s have parents and children. This relationship allows an
+  /// entire ancestor path to "have focus", but only the lowest level descendant
+  /// in that path has "primary focus". This path is important because various
+  /// widgets alter their presentation or behavior based on whether or not they
+  /// currently have focus, even if they only have "non-primary focus".
+  ///
+  /// When the popover is visible, it will have primary focus.
+  /// Moreover, because the popover is built in an `Overlay`, none of your
+  /// widgets are in the natural focus path for that popover. Therefore, if you
+  /// need your widget tree to retain focus while the popover is visible, then
+  /// you need to provide the [FocusNode] that the popover should use as its
+  /// parent, thereby retaining focus for your widgets.
+  final FocusNode? parentFocusNode;
+
+  /// A [GlobalKey] to a widget that determines the bounds where the popover can be displayed.
+  ///
+  /// Passing a [boundaryKey] causes the popover to be confined to the bounds of the widget
+  /// bound to the [boundaryKey].
+  ///
+  /// If `null`, the popover is confined to the screen bounds, defined by the result of `MediaQuery.sizeOf`.
+  final GlobalKey? boundaryKey;
+
+  /// Called when the user taps outside of the popover.
+  ///
+  /// If `null`, tapping outside closes the popover.
+  final void Function(PopoverController) onTapOutside;
```

Comment:
Can we add the `tapRegionGroupId` that I added in the super_editor PR here?  

I added the property and the test in the following PR: https://github.com/superlistapp/super_editor/pull/1819